button-merchant/src/main/java/com/usebutton/merchant/ButtonApiImpl.java

```diff
@@ -166,20 +183,47 @@ public PostInstallLink getPendingLink(String applicationId, String ifa,
         return null;
     }
 
-    private void initializeUrlConnection(HttpURLConnection urlConnection) throws ProtocolException {
+    private void initializeUrlConnection(HttpsURLConnection urlConnection) throws
+            CertificateException, KeyStoreException, NoSuchAlgorithmException, IOException,
+            KeyManagementException {
+
         urlConnection.setConnectTimeout(CONNECT_TIMEOUT);
         urlConnection.setReadTimeout(READ_TIMEOUT);
         urlConnection.setRequestProperty("User-Agent", getUserAgent());
         urlConnection.setRequestProperty("Accept", CONTENT_TYPE_JSON);
         urlConnection.setRequestProperty("Content-Type", CONTENT_TYPE_JSON);
         urlConnection.setRequestMethod("POST");
         urlConnection.setDoOutput(true);
+
+        // Public key pinning is only available API 17 and above
+        // Default to CA pinning for older Android versions
+        if (Build.VERSION.SDK_INT > Build.VERSION_CODES.JELLY_BEAN_MR1) {
+            Set<String> keys = sslManager.getCertificateProvider().getPublicKeys();
+
+            TrustManagerFactory trustManagerFactory =
+                    TrustManagerFactory.getInstance(TrustManagerFactory.getDefaultAlgorithm());
+            trustManagerFactory.init((KeyStore) null);
+            X509TrustManager trustManager = null;
+            for (TrustManager tm : trustManagerFactory.getTrustManagers()) {
+                if (tm instanceof X509TrustManager) {
+                    trustManager = (X509TrustManager) tm;
+                    break;
+                }
+            }
+            X509TrustManagerExtensions extension = new X509TrustManagerExtensions(trustManager);
+            Encoder encoder = new AndroidEncoder();
+
+            SSLUtils.validatePinning(extension, encoder, urlConnection, keys);
```

Comment:
Currently there does not seem to be any test coverage to verify public key pinning is happening. If I comment out this line then all tests still pass.

I think we need a test that overrides `Build.VERSION.SDK_INT` to be simulate API 17+.